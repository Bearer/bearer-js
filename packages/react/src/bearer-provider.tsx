import kebabCase from 'lodash.kebabcase'
import * as React from 'react'
import bearer, { BearerInstance } from '@bearer/js'

// TODO: add JSDoc
interface IBearerProviderProps {
  clientId: string
  domObserver?: boolean
  integrationHost?: string
  initialContext?: any
  onUpdate?(currentState: any): void
}

export interface IBearerContextValue {
  bearer: BearerInstance
  state: any
  handlePropUpdates?(e: any): void
}

export const BearerContext = React.createContext<IBearerContextValue>({
  bearer: bearer.instance!,
  state: {},
  handlePropUpdates: () => {}
})

interface IBearerProviderState {
  bearer: BearerInstance
  integrationState: any
}

export default class BearerProvider extends React.Component<IBearerProviderProps, IBearerProviderState> {
  constructor(props: IBearerProviderProps) {
    super(props)
    this.state = {
      bearer: this.initBearerFromProps(props),
      integrationState: props.initialContext || {}
    }
  }

  public render() {
    const value = {
      bearer: this.state.bearer,
      handlePropUpdates: this.handlePropUpdates,
      state: this.state.integrationState
    }
    return <BearerContext.Provider value={value}>{this.props.children}</BearerContext.Provider>
  }

  componentWillReceiveProps(newProps: IBearerProviderProps, oldProps: IBearerProviderProps) {
    if (newProps.clientId !== oldProps.clientId) {
      this.setState(state => ({ ...state, bearer: this.initBearerFromProps(newProps) }))
    }
  }

  public componentDidUpdate(_prevProps: IBearerProviderProps, prevState: any) {
    if (this.props.onUpdate && this.state !== prevState) {
      this.props.onUpdate(this.state)
    }
  }

  private readonly handlePropUpdates = (e: CustomEvent) => {
    e.stopPropagation()
    const payload = e.detail
    const kebabPayload = Object.keys(payload).reduce((acc: any, key: string) => {
      acc[kebabCase(key)] = payload[key]
      return acc
    }, {})
    this.setState(state => ({
      ...state,
      integrationState: {
        ...state.integrationState,
        ...kebabPayload
      }
    }))
  }

  private initBearerFromProps(props: IBearerProviderProps) {
    const { integrationHost, domObserver } = props
    return bearer(props.clientId, { integrationHost, domObserver })
  }
}
