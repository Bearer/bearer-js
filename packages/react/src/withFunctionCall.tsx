import * as React from 'react'

import { BearerContext } from './bearer-provider'

export interface WithFetchProps<TReturnedData> {
  loading: boolean
  error?: any
  data?: TReturnedData
  fetch: (params: any) => void
}

export interface IState<TReturnedData> {
  loading: boolean
  data?: TReturnedData
  error?: any
}

export const withFunctionCall = function<TReturnedData, TP extends object = {}>(
  integrationId: string,
  functionName: string
) {
  return (WrappedComponent: React.ComponentType<TP & WithFetchProps<TReturnedData>>) =>
    class FetcherComponent extends React.Component<TP, IState<TReturnedData>> {
      static displayName = `withFunctionCall(${functionName})(${getDisplayName(WrappedComponent)})`
      static contextType = BearerContext
      context!: React.ContextType<typeof BearerContext>

      constructor(props: TP) {
        super(props)
        this.state = {
          loading: false
        }
      }

      fetch = (params: any = {}) => {
        this.setState({ ...this.state, error: null, loading: true })
        this.context.bearer
          .functionFetch(integrationId, functionName, params)
          .then(({ data }) => {
            this.setState({ ...this.state, data })
          })
          .catch(({ error }) => {
            this.setState({ ...this.state, error })
          })
          .then(() => {
            this.setState({ ...this.state, loading: false })
          })
      }

      public render() {
        return (
          <WrappedComponent
            {...this.props}
            error={this.state.error}
            data={this.state.data}
            fetch={this.fetch}
            loading={this.state.loading}
          />
        )
      }
    }
}

function getDisplayName(WrappedComponent: any) {
  return WrappedComponent.displayName || WrappedComponent.name || 'Component'
}
