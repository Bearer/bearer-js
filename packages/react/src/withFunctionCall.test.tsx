import React, { createContext } from 'react'
import * as Renderer from 'react-test-renderer'
import * as ShallowRenderer from 'react-test-renderer/shallow'
import { BearerContext } from './bearer-provider'
import { withFunctionCall } from './withFunctionCall'

interface TProps {
  loading: boolean
  error?: any
  data?: { title: string }
  fetch: (params: any) => void
  sommething: string
}

class DummyWithFetchDataComponent extends React.Component<TProps> {
  render() {
    if (this.props.loading) {
      return <div>Loading</div>
    }

    if (this.props.error) {
      return <div>Error: {this.props.error}</div>
    }

    if (this.props.data) {
      return <div>Some data passed: {this.props.data}</div>
    }

    return (
      <div>
        <button onClick={this.props.fetch}>Click to fetch</button>
      </div>
    )
  }
}

describe('withFunctionCall', () => {
  it('exports a function', () => {
    expect(withFunctionCall).toBeInstanceOf(Function)
    expect(withFunctionCall).toHaveLength(2)
  })

  describe('Wrapped Component', () => {
    const renderer = ShallowRenderer.createRenderer()
    const WithFetch = withFunctionCall<{ title: string }, { sommething: string }>('integrationName', 'GimmeData')(
      DummyWithFetchDataComponent
    )

    it('renders the wrapped component adn forward props', () => {
      const tree = renderer.render(<WithFetch sommething="ok" />)

      expect(tree).toMatchSnapshot()
    })

    it('sets the display name', () => {
      expect(WithFetch.displayName).toEqual('withFunctionCall(GimmeData)(DummyWithFetchDataComponent)')
    })
  })

  describe('fetch behaviours', () => {
    const WithFetch = withFunctionCall<{ title: string }, any>('integrationName', 'GimmeData')(
      DummyWithFetchDataComponent
    )

    function render(bearer: any) {
      return Renderer.create(
        <BearerContext.Provider value={{ bearer, state: {} }}>
          <WithFetch />
        </BearerContext.Provider>
      )
    }

    it('forwards loading on fetch', () => {
      const bearer = {
        functionFetch: jest.fn(() => new Promise(() => {}))
      } as any

      const rendered = render(bearer)

      rendered.root.findByType('button').props.onClick()
      expect(rendered.toJSON()).toMatchSnapshot()
    })

    it('forwards data when success', async () => {
      const bearer = {
        functionFetch: jest.fn(() => Promise.resolve({ data: 'Sponge bob is alive' }))
      } as any

      const rendered = render(bearer)

      rendered.root.findByType('button').props.onClick()
      // we delay expectations ;-)
      await bearer.functionFetch()

      expect(rendered.toJSON()).toMatchSnapshot()
    })

    it('forwards data when failure', async () => {
      const bearer = {
        functionFetch: jest.fn(() => Promise.reject({ error: 'data error' }))
      } as any

      const rendered = render(bearer)

      rendered.root.findByType('button').props.onClick()
      // we delay expectations ;-)
      try {
        await bearer.functionFetch()
      } catch (_e) {}

      expect(rendered.toJSON()).toMatchSnapshot()
    })

    describe('callbacks', () => {
      // TODO
    })
  })
})
