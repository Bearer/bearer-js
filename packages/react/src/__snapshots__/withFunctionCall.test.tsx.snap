// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withFunctionCall Wrapped Component renders the wrapped component adn forward props 1`] = `
<DummyWithFetchDataComponent
  fetch={[Function]}
  loading={false}
  sommething="ok"
/>
`;

exports[`withFunctionCall fetch behaviours forwards data when failure 1`] = `
<div>
  Error: 
  data error
</div>
`;

exports[`withFunctionCall fetch behaviours forwards data when success 1`] = `
<div>
  Some data passed: 
  Sponge bob is alive
</div>
`;

exports[`withFunctionCall fetch behaviours forwards loading on fetch 1`] = `
<div>
  Loading
</div>
`;
